nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java

```diff
@@ -107,110 +173,212 @@ public NullnessHint onDataflowVisitMethodInvocation(
+      }
+      if (!supported) {
+        // Too many arguments involved, or unsupported @Contract features. On to next clause in
+        // the
+        // contract expression
+        continue;
+      }
+      if (arg == null) {
+        // The antecedent is unconditionally true. Check for the ... -> !null case and set the
+        // return nullness accordingly
+        if (consequent.equals("!null")) {
+          return NullnessHint.FORCE_NONNULL;
         }
+        continue;
+      }
+      if (argAntecedentNullness == null) {
+        throw new IllegalStateException("argAntecedentNullness should have been set");
+      }
+      // The nullness of one argument is all that matters for the antecedent, let's negate the
+      // consequent to fix the nullness of this argument.
+      AccessPath accessPath = AccessPath.getAccessPathForNodeNoMapGet(arg, apContext);
+      if (accessPath == null) {
+        continue;
+      }
+      if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NULL)) {
+        // If arg being null implies the return of the method being false, then the return
+        // being true implies arg is not null and we must mark it as such in the then update.
+        thenUpdates.set(accessPath, Nullness.NONNULL);
+      } else if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // If arg being non-null implies the return of the method being false, then the return
+        // being true implies arg may be null and we must mark it as such in the then update.
+        thenUpdates.set(accessPath, Nullness.NULLABLE);
+      } else if (consequent.equals("true") && argAntecedentNullness.equals(Nullness.NULL)) {
+        // If arg being null implies the return of the method being true, then the return being
+        // false implies arg is not null and we must mark it as such in the else update.
+        elseUpdates.set(accessPath, Nullness.NONNULL);
+      } else if (consequent.equals("true") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // If arg being non-null implies the return of the method being true, then the return being
+        // false implies arg may be null and we must mark it as such in the else update.
+        elseUpdates.set(accessPath, Nullness.NULLABLE);
+      } else if (consequent.equals("fail") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // Arg being non-null implies the method throws an exception, then we can mark it as
+        // null on both non-exceptional exits from the method.
+        bothUpdates.set(accessPath, Nullness.NULL);
+      } else if (consequent.equals("fail") && argAntecedentNullness.equals(Nullness.NULL)) {
+        // If arg being null implies the method throws an exception, then we can mark it as
+        // non-null on both non-exceptional exits from the method
+        bothUpdates.set(accessPath, Nullness.NONNULL);
       }
     }
     return NullnessHint.UNKNOWN;
   }
```

Comment:
As a meta-comment: this method grows increasingly long and it's basically a maze of cases, some refactoring for readability might make sense, either now or as a follow up...

Reply:
Agreed, I'd tried to avoid modification where possible, but at this point I think I've slipped into the worst case middle-ground between refactoring the existing code and introducing new behavior